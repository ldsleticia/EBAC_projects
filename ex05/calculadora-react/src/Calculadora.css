body {
  margin: 0;
  font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", "Roboto", "Oxygen",
    "Ubuntu", "Cantarell", "Fira Sans", "Droid Sans", "Helvetica Neue",
    sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
}

code {
  font-family: source-code-pro, Menlo, Monaco, Consolas, "Courier New",
    monospace;
}

.Calculadora {
  font-family: sans-serif;
  text-align: center;
  border: 2px solid #000;
  border-radius: 5px;
  width: 15rem;
  margin-left: 50em;
  margin-top: 10em;
  padding-bottom: 5em;
}

.numberButtons,
.operatorButtons {
  background-color: #b57fff;
  color: #ffffff;
  border: none;
  cursor: pointer;
  transition: background-color 0.3s;
}

.numberButtons:hover,
.operatorButtons:hover {
  background-color: #9868f5;
}

.result {
  color: #7c51ea;
  text-align: right;
}
