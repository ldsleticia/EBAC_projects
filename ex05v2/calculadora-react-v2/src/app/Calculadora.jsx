"use client";
import "../app/calculadora.css";
import React, { useEffect, useState } from "react";
export default function Calculadora() {
  const [numberDigit, setNumberDigit] = useState("");
  const [isResult, setIsResult] = useState(false);
  const [lastCharIsOperator, setLastCharIsOperator] = useState(false);

  useEffect(() => {
    window.addEventListener("keydown", handleKeyDown);

    return () => {
      window.removeEventListener("keydown", handleKeyDown);
    };
  }, []);

  const handleKeyDown = (event) => {
    console.log(event.key);
    console.log(numberDigit);
    if (/^\d$/.test(event.key)) {
      handleAddNumber(event.key);
      console.log("estou no if da handleAddNumber");
      console.log("o number digit é", numberDigit);
    } else {
      if (/^[+\-*\/=Cc\n\r]$/.test(event.key)) {
        handleAddSimbol(event.key);
      }
      if (event.key === "Enter" || event.key === "Return") {
        evaluateExpression(numberDigit);
      }
    }
  };

  function handleAddNumber(number) {
    if (numberDigit && !isResult) {
      setNumberDigit((prevNumberDigit) => prevNumberDigit.concat(number));
    } else {
      setNumberDigit(number);
      setIsResult(false);
    }
  }

  function evaluateExpression(expression) {
    const regex = /^\d+[\+\-\*\/]\d+$/;
    const result = eval(expression);

    if (!lastCharIsOperator && !regex.test(expression)) {
      setNumberDigit(
        "Expressão inválida. Limpe a calculadora e refaça a expressão"
      );
      setIsResult(true);
      return;
    }

    if (expression.includes("/0")) {
      setNumberDigit("Divisão por zero não é permitida");
      setIsResult(true);
      return;
    } else {
      setNumberDigit(result.toString());
      setIsResult(false);
      console.log(result);
    }
  }

  function handleAddSimbol(simbol) {
    if (simbol === "=") {
      console.log(numberDigit);
      evaluateExpression(numberDigit);
    } else if (simbol === "C" || simbol === "c") {
      setNumberDigit("");
      setIsResult(false);
    } else if (numberDigit) {
      setNumberDigit((prevNumberDigit) => prevNumberDigit + simbol);
      setIsResult(false);
      setLastCharIsOperator(/^[+\-*\/]$/.test(simbol));
    }
  }

  const numbers = ["7", "8", "9", "4", "5", "6", "1", "2", "3", "0"];
  const simbols = ["/", "*", "+", "-"];
  return (
    <main className="mainClass">
      <h1>{numberDigit}</h1>
      <div className="container">
        {numbers.map((number) => (
          <button
            className="item"
            onClick={() => handleAddNumber(number)}
            key={number}
          >
            {number}
          </button>
        ))}
        {simbols.map((simbol) => (
          <button
            className="simbols"
            onClick={() => {
              handleAddSimbol(simbol);
            }}
            key={simbol}
          >
            {simbol}
          </button>
        ))}
        <button
          className="ac-button"
          onClick={() => {
            handleAddSimbol("C");
          }}
        >
          {" "}
          C{" "}
        </button>
        <button
          className="equal-button"
          onClick={() => {
            handleAddSimbol("=");
          }}
        >
          {" "}
          ={" "}
        </button>
      </div>
    </main>
  );
}
