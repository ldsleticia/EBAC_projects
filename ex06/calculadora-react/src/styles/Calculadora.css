body {
  background-image: url(https://img.freepik.com/fotos-gratis/respingo-colorido-abstrato-3d-background-generativo-ai-background_60438-2509.jpg);
  background-position: center;
}

.mainClass {
  display: grid;
  grid-template-columns: 1fr;
  grid-template-rows: auto 1fr;
  align-content: center;
  justify-items: center;
}

.container {
  display: grid;
  grid-template-columns: repeat(4, 1fr);
  gap: 10px;
  border-radius: 10px;
  background-color: #c6dabf;
}

h1 {
  grid-column: 1;
  grid-row: 1;
  text-align: center;
  color: #000000;
}
