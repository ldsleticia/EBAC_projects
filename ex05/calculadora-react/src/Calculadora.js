import React, { Component } from "react";
import "./Calculadora.css"

class Calculadora extends Component {
  constructor() {
    super();

    this.state = {
      numberDigit: "",
      simbolDigit: "",
      isResult: false,
      lastCharIsOperator: false,
    };
  }

  componentDidMount() {
    window.addEventListener("keydown", this.handleKeyDown);
  }

  componentWillUnmount() {
    window.removeEventListener("keydown", this.handleKeyDown);
  }

  handleKeyDown = (event) => {
    if (/^\d$/.test(event.key)) {
      this.handleAddNumber(event.key);
    } else {
      if (/^[+\-*\/=Cc\n\r]$/.test(event.key)) {
        this.handleAddSimbol(event.key);
      }
      if (event.key === "Enter" || event.key === "Return") {
        this.evaluateExpression(this.state.numberDigit);
      }
    }
  };

  evaluateExpression(expression) {
    const regex = /^\d+[\+\-\*\/]\d+$/;

    if (this.state.lastCharIsOperator) {
      this.setState({
        numberDigit:
          "Expressão inválida. Limpe a calculadora e refaça a expressão",
        isResult: true,
      });
    }

    if (regex.test(expression)) {
      const result = eval(expression);
      this.setState({
        numberDigit: result.toString(),
        isResult: false,
      });
    } else {
      return;
    }
  }

  handleAddSimbol(simbol) {
    if (simbol === "=" || simbol === "Enter" || simbol === "Return") {
      this.evaluateExpression(this.state.numberDigit);
    } else if (simbol === "C" || simbol === "c") {
      this.setState({
        numberDigit: "",
        isResult: false,
      });
    } else if (this.state.numberDigit) {
      this.setState({
        numberDigit: this.state.numberDigit.concat(simbol),
        isResult: false,
        lastCharIsOperator: /^[+\-*\/]$/.test(simbol),
      });
    }
  }

  handleAddNumber(number) {
    if (this.state.numberDigit && !this.state.isResult) {
      this.setState({
        numberDigit: this.state.numberDigit.concat(number),
      });
    } else {
      this.setState({
        numberDigit: number,
        isResult: false,
      });
    }
  }

  render() {
    const numbers = ["0", "1", "2", "3", "4", "5", "6", "7", "8", "9"];
    const simbols = ["+", "=", "*", "-", "/", "C"];

    return (
      <div className="Calculadora">
        {/* exibir o resultado em um componente */}
        <h1 className="result">{this.state.numberDigit ? this.state.numberDigit : "0"}</h1>
        <section>
          {numbers.map((number) => (
            <button className="numberButtons" onClick={() => this.handleAddNumber(number)} key={number}>
              {number}
            </button>
          ))}
          </section>
          <section>
            {simbols.map((simbol) => (
              <button className="operatorButtons" onClick={() => this.handleAddSimbol(simbol)} key={simbol}>
                {simbol}
              </button>
          ))}
          </section>
          
        
      </div>
    );
  }
}

export default Calculadora;
