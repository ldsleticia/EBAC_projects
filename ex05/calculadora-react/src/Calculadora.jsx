import React, { Component } from "react";
import "./Calculadora.css"

class Calculadora extends Component {
  constructor() {
    super();

    this.state = {
      numberDigit: "",
      isResult: false,
      lastCharIsOperator: false,
    };
  }

  componentDidMount() {
    window.addEventListener("keydown", this.handleKeyDown);
  }

  componentWillUnmount() {
    window.removeEventListener("keydown", this.handleKeyDown);
  }

  handleKeyDown = (event) => {
    if (/^\d$/.test(event.key)) {
      this.handleAddNumber(event.key);
    } else {
      if (/^[+\-*\/=Cc\n\r]$/.test(event.key)) {
        this.handleAddSimbol(event.key);
      }
      if (event.key === "Enter" || event.key === "Return") {
        this.evaluateExpression(this.state.numberDigit);
      }
    }
  };

  evaluateExpression(expression) {
    const regex = /^\d+[\+\-\*\/]\d+$/;

    if (this.state.lastCharIsOperator) {
      this.setState({
        numberDigit:
          "Expressão inválida. Limpe a calculadora e refaça a expressão",
        isResult: true,
      });
    }

    if (regex.test(expression)) {
      const result = eval(expression);
      this.setState({
        numberDigit: result.toString(),
        isResult: false,
      });
    } else {
      return;
    }
  }

  handleAddSimbol(simbol) {
    if (simbol === "=" || simbol === "Enter" || simbol === "Return") {
      this.evaluateExpression(this.state.numberDigit);
    } else if (simbol === "C" || simbol === "c") {
      this.clearResult();
    } else if (this.state.numberDigit) {
      this.setState({
        numberDigit: this.state.numberDigit.concat(simbol),
        isResult: false,
        lastCharIsOperator: /^[+\-*\/]$/.test(simbol),
      });
    }
  }

  clearResult() {
    this.setState({
      numberDigit: "",
      isResult: false,
    });
  }

  handleAddNumber(number) {
    if (this.state.numberDigit && !this.state.isResult) {
      this.setState({
        numberDigit: this.state.numberDigit.concat(number),
      });
    } else {
      this.setState({
        numberDigit: number,
        isResult: false,
      });
    }
  }

  render() {
    const numbers = ["0", "1", "2", "3", "4", "5", "6", "7", "8", "9"];
    const simbols = ["+", "*", "-", "/"];

    return (
      <main className="mainClass">
      <h1>{this.state.numberDigit ? this.state.numberDigit : "0"}</h1>
      <div className="container">
          {numbers.map((number) => (
            <button className="item" onClick={() => this.handleAddNumber(number)} key={number}>
              {number}
            </button>
          ))}
            {simbols.map((simbol) => (
              <button className="simbols" onClick={() => this.handleAddSimbol(simbol)} key={simbol}>
                {simbol}
              </button>
          ))}
          <button
          className="ac-button"
          onClick={() => {
            this.handleAddSimbol("C");
          }}
        >
          {" "}
          C{" "}
        </button>
        <button
          className="equal-button"
          onClick={() => {
            this.handleAddSimbol("=");
          }}
        >
          {" "}
          ={" "}
        </button>
      </div>
      </main>
    );
  }
}

export default Calculadora;
