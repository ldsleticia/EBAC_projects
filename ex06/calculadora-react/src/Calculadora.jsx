import React, { Component } from "react";
import "../src/styles/Calculadora.css";
import Resultado from "./Resultado";
import Teclado from "./Teclado";

class Calculadora extends Component {
  constructor() {
    super();

    this.state = {
      numberDigit: "",
      isResult: false,
      lastCharIsOperator: false,
    };

    this.handleAddNumber = this.handleAddNumber.bind(this);
    this.handleAddSimbol = this.handleAddSimbol.bind(this);
  }

  componentDidMount() {
    window.addEventListener("keydown", this.handleKeyDown);
  }

  componentWillUnmount() {
    window.removeEventListener("keydown", this.handleKeyDown);
  }

  handleKeyDown = (event) => {
    if (event.key === "Backspace" || event.key === "Delete") {
      const updatedNumber = this.state.numberDigit.slice(0, -1);
      this.setState({
        numberDigit: updatedNumber,
      });
    }
    if (/^\d$/.test(event.key)) {
      this.handleAddNumber(event.key);
    } else {
      if (/^[+\-*\/=Cc\n\r]$/.test(event.key)) {
        this.handleAddSimbol(event.key);
      }
      if (event.key === "Enter" || event.key === "Return") {
        this.evaluateExpression(this.state.numberDigit);
      }
    }
  };

  evaluateExpression(expression) {
    const regex = /^\d+[\+\-\*\/]\d+$/;

    if (this.state.lastCharIsOperator) {
      this.setState({
        numberDigit:
          "Expressão inválida. Limpe a calculadora e refaça a expressão",
        isResult: true,
      });
    }

    if (regex.test(expression)) {
      const result = eval(expression);
      this.setState({
        numberDigit: result.toString(),
        isResult: true,
      });
    } else {
      return;
    }
  }

  handleAddSimbol(simbol) {
    if (simbol === "=" || simbol === "Enter" || simbol === "Return") {
      this.evaluateExpression(this.state.numberDigit);
      this.setState({
        isResult: true,
      });
      if(this.state.isResult) {
        this.clearResult();
      }
    } 
    else if (simbol === "C" || simbol === "c") {
      this.clearResult();
    } else if (this.state.numberDigit) {
      this.setState({
        numberDigit: this.state.numberDigit.concat(simbol),
        isResult: false,
        lastCharIsOperator: /^[+\-*\/]$/.test(simbol),
      });
    }
  }

  clearResult() {
    this.setState({
      numberDigit: "",
      isResult: false,
    });
  }

  handleAddNumber(number) {
    if (this.state.numberDigit && !this.state.isResult) {
      this.setState({
        numberDigit: this.state.numberDigit.concat(number),
      });
    } else {
      this.setState({
        numberDigit: number,
        isResult: false,
      });
    }
  }

  render() {
    const numbers = ["0", "1", "2", "3", "4", "5", "6", "7", "8", "9"];
    const simbols = ["+", "*", "-", "/"];

    return (
      <main className="mainClass">
        <div className="box">
          <Resultado numberDigit={this.state.numberDigit} />
        </div>
        <div className="overlay">
          <Teclado
            numbers={numbers}
            simbols={simbols}
            handleAddNumber={this.handleAddNumber}
            handleAddSimbol={this.handleAddSimbol}
          />
        </div>
      </main>
    );
  }
}

export default Calculadora;
