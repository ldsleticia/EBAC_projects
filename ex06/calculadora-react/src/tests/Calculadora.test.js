import React from "react";
import { shallow } from "enzyme";
import Calculadora from "../Calculadora";

describe("Calculadora", () => {
  it("Deve renderizar sem erros", () => {
    const wrapper = shallow(<Calculadora />);
    expect(wrapper).toMatchSnapshot();
  });
  it("Deve inicializar com o estado correto", () => {
    const wrapper = shallow(<Calculadora />);
    expect(wrapper.state("numberDigit")).toBe("");
    expect(wrapper.state("isResult")).toBe(false);
    expect(wrapper.state("lastCharIsOperator")).toBe(false);
  });
  it("Deve adicionar números na tela", () => {
    const wrapper = shallow(<Calculadora />);
    wrapper.instance().handleAddNumber("1");
    expect(wrapper.state("numberDigit")).toBe("1");
    wrapper.instance().handleAddNumber("2");
    expect(wrapper.state("numberDigit")).toBe("12");
  });
  it("Deve adicionar símbolo de + na tela", () => {
    const wrapper = shallow(<Calculadora />);
    wrapper.instance().handleAddNumber("1");
    expect(wrapper.state("numberDigit")).toBe("1");
    wrapper.instance().handleAddSimbol("+");
    expect(wrapper.state("numberDigit")).toBe("1+");
  });
  it("Deve adicionar um sinal de - na tela", () => {
    const wrapper = shallow(<Calculadora />);
    wrapper.instance().handleAddNumber("1");
    expect(wrapper.state("numberDigit")).toBe("1");
    wrapper.instance().handleAddSimbol("-");
    expect(wrapper.state("numberDigit")).toBe("1-");
  });
  it("Deve adicionar um sinal de - na tela", () => {
    const wrapper = shallow(<Calculadora />);
    wrapper.instance().handleAddNumber("1");
    expect(wrapper.state("numberDigit")).toBe("1");
    wrapper.instance().handleAddSimbol("*");
    expect(wrapper.state("numberDigit")).toBe("1*");
  });
  it("Deve adicionar um sinal de - na tela", () => {
    const wrapper = shallow(<Calculadora />);
    wrapper.instance().handleAddNumber("1");
    expect(wrapper.state("numberDigit")).toBe("1");
    wrapper.instance().handleAddSimbol("/");
    expect(wrapper.state("numberDigit")).toBe("1/");
  });
  it('Deve calcular o resultado quando o símbolo é "="', () => {
    const wrapper = shallow(<Calculadora />);
    wrapper.instance().setState({ numberDigit: "2+3" });
    wrapper.instance().handleAddSimbol("=");
    wrapper.instance().evaluateExpression("2+3");
    expect(wrapper.instance().state.numberDigit).toBe("5");
    expect(wrapper.instance().state.isResult).toBe(true);
  });
});
