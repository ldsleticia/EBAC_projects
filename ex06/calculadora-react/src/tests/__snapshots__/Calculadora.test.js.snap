// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calculadora Deve renderizar sem erros 1`] = `ShallowWrapper {}`;

exports[`Calculadora should render without errors 1`] = `ShallowWrapper {}`;
